test_ask.py

```diff
@@ -82,52 +82,236 @@ def test_maybe_place_space_before(self):
+		test_value = 'parsed string\t'
+		expected = '\t'
+		self.assertEqual(expected, ask.get_current_tab_level(test_value))
+
+	def test_get_current_tab_level_tab_in_string(self):
+		test_value = 'parsed \t string'
+		expected = '\t'
+		self.assertEqual(expected, ask.get_current_tab_level(test_value))
+
+	def test_get_current_tab_level_newline(self):
+		test_value = '\n'
+		expected = ''
+		self.assertEqual(expected, ask.get_current_tab_level(test_value))
+
+	def test_get_current_tab_level_newline_then_tab(self):
+		test_value = '\n\t'
+		expected = '\t'
+		self.assertEqual(expected, ask.get_current_tab_level(test_value))
+		
+	def test_get_current_tab_level_tab_then_newline_then_tab(self):
+		test_value = 'parsed \t string \n \t'
+		expected = '\t'
+		self.assertEqual(expected, ask.get_current_tab_level(test_value))
+
+	# tokens_grouped_by_lines
+	def test_tokens_grouped_by_lines(self):
+		tokens = [
+			['VAR', 'variable'],
+			['FORMAT', '\n'],
+			['VAR', 'variable'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['VAR', 'variable'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['FORMAT', '\t'],
+			['VAR', 'variable'],
+		]
+
+		excpected = [
+			[
+				['VAR', 'variable'],
+			],
+			[
+				['FORMAT', '\n'],
+				['VAR', 'variable'],
+			],
+			[
+				['FORMAT', '\n'],
+				['FORMAT', '\t'],
+				['VAR', 'variable'],
+			],
+			[
+				['FORMAT', '\n'],
+				['FORMAT', '\t'],
+				['FORMAT', '\t'],
+				['VAR', 'variable'],
+			],
+		]
+
+		self.assertEqual(excpected, ask.tokens_grouped_by_lines(tokens))
+
+	# insert_indentation_group_markers
+	def test_insert_indention_group_markers(self):
+		tokens = [
+			['FUNC_DEF', 'function'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['FUNC', 'print'],
+			['STR', 'text'],
+			['OP', ')'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['KEYWORD', 'if'],
+			['VAR', 'var'],
+			['OP', ':'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['FORMAT', '\t'],
+			['FUNC', 'print'],
+			['STR', 'text'],
+			['OP', ')'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['FUNC', 'print'],
+			['STR', 'text'],
+			['OP', ')'],
+		]
+
+		expected = [
+			['GROUP', 'start'],
+			['FUNC_DEF', 'function'],
+			['GROUP', 'start'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['FUNC', 'print'],
+			['STR', 'text'],
+			['OP', ')'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['KEYWORD', 'if'],
+			['VAR', 'var'],
+			['OP', ':'],
+			['GROUP', 'start'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['FORMAT', '\t'],
+			['FUNC', 'print'],
+			['STR', 'text'],
+			['OP', ')'],
+			['GROUP', 'end'],
+			['FORMAT', '\n'],
+			['FORMAT', '\t'],
+			['FUNC', 'print'],
+			['STR', 'text'],
+			['OP', ')'],
+			['GROUP', 'end'],
+			['GROUP', 'end']
+		]
+
+		self.assertEqual(expected, ask.insert_indention_group_markers(tokens))
+
+	# add_part
+	def test_add_part_string_true_code_with_newline(self):
+		is_string = True
+		code = 'code \n'
+		expected = (
+			[{'is_string': True, 'code': 'code \n'}], 
```

Comment:
`add_part()` will always return `code` the same way as it gets passed in, so no need to hardcode `'code': 'code \n'` when you have already defined `code` on line 278. Do this instead: `'code': code` Do this in all of `add_part()`'s tests.

Reply:
done.